cartography/intel/azure/sql.py

```diff
@@ -0,0 +1,742 @@
+
+    ingest_server = """
+    MERGE (s:AzureServer{id: {ServerId}})
+    ON CREATE SET s.firstseen = timestamp(), 
+    s.id = {ServerId}, s.name = {Name}, 
+    s.resourcegroup = {ResourceGroup}, s.location = {Location}
+    SET s.lastupdated = {azure_update_tag}, 
+    s.kind = {Kind},
+    s.state = {State},
+    s.version = {Version}
+    WITH s
+    MATCH (owner:AzureAccount{id: {AZURE_SUBSCRIPTION_ID}})
+    MERGE (owner)-[r:RESOURCE]->(s)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {azure_update_tag}
+    """
+
+    for server in server_list:
+        data = server.as_dict()
+        neo4j_session.run(
+            ingest_server,
+            ServerId=data['id'],
+            ResourceGroup=data['resourceGroup'],
+            Name=data['name'],
+            Location=data['location'],
+            Kind=data['kind'],
+            State=data['properties']['state'],
+            Version=data['properties']['version'],
+            AZURE_SUBSCRIPTION_ID=subscription_id,
+            azure_update_tag=azure_update_tag,
+        )
+
+
+@timeit
+def sync_server_details(neo4j_session, credentials, subscription_id, server_list, sync_tag):
+    details = get_server_details(credentials, subscription_id, server_list)
+    load_server_details(neo4j_session, credentials, subscription_id, details, sync_tag)
+
+
+@timeit
+def get_server_details(credentials, subscription_id, server_list):
+    for server in server_list:
+        dns_alias = get_dns_aliases(credentials, subscription_id, server)
+        ad_admins = get_ad_admins(credentials, subscription_id, server)
+        recoverable_databases = get_recoverable_databases(credentials, subscription_id, server)
+        restorable_dropped_databases = get_restorable_dropped_databases(credentials, subscription_id, server)
+        failover_groups = get_failover_groups(credentials, subscription_id, server)
+        elastic_pools = get_elastic_pools(credentials, subscription_id, server)
+        databases = get_databases(credentials, subscription_id, server)
+        yield server['id'], server['name'], server['resourceGroup'], dns_alias, ad_admins, recoverable_databases, restorable_dropped_databases, failover_groups, elastic_pools, databases
+
+
+@timeit
+def get_dns_aliases(credentials, subscription_id, server):
+    try:
+        client = get_client(credentials, subscription_id)
+        dns_aliases = list(client.server_dns_aliases.list_by_server(server['resourceGroup'], server['name']))
+
+    except Exception as e:
+        print(f'Failed to retrieve server DNS Aliases : {e}')
+        return []
+
+    return dns_aliases
+
+
+@timeit
+def get_ad_admins(credentials, subscription_id, server):
+    try:
+        client = get_client(credentials, subscription_id)
+        ad_admins = list(client.server_azure_ad_administrators.list_by_server(server['resourceGroup'], server['name']))
+
+    except Exception as e:
+        print(f'Failed to retrieve server azure AD Administrators : {e}')
```

Comment:
same as above, let's use logger for printing exception

Reply:
done